### DISCUSSION TITLE
Use of new go proxy server

### DISCUSSION THREAD

**Discussion by brianmay** [teslamate-org/teslamate#3527] [General]
Just some notes of what I have done so far:

I use nix:

```sh
$ nix-shell -p go -p openssl
```

Download, compile go program:

```sh
[nix-shell:~/tree/3rdparty/vehicle-command]$  git clone https://github.com/teslamotors/vehicle-command.git
Cloning into 'vehicle-command'...
remote: Enumerating objects: 205, done.
remote: Counting objects: 100% (40/40), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 205 (delta 16), reused 16 (delta 10), pack-reused 165
Receiving objects: 100% (205/205), 226.57 KiB | 3.19 MiB/s, done.
Resolving deltas: 100% (35/35), done.
[nix-shell:~/tree/3rdparty/vehicle-command]$ go get ./...
go: downloading github.com/99designs/keyring v1.2.2
go: downloading golang.org/x/term v0.5.0
go: downloading github.com/go-ble/ble v0.0.0-20220207185428-60d1eecf2633
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading golang.org/x/sys v0.8.0
go: downloading github.com/dvsekhvalnov/jose2go v1.5.0
go: downloading github.com/gsterjov/go-libsecret v0.0.0-20161001094733-a6f4afe4910c
go: downloading github.com/mtibben/percent v0.2.1
go: downloading github.com/99designs/go-keychain v0.0.0-20191008050251-8e49817e8af4
go: downloading github.com/danieljoos/wincred v1.2.0
go: downloading github.com/godbus/dbus v0.0.0-20190726142602-4481cbc300e2
go: downloading github.com/raff/goble v0.0.0-20190909174656-72afc67d6a99
go: downloading github.com/tinygo-org/cbgo v0.0.4
go: downloading github.com/mgutz/logxi v0.0.0-20161027140823-aebf8a7d67ab
go: downloading github.com/pkg/errors v0.8.1
go: downloading github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
go: downloading github.com/mattn/go-isatty v0.0.12
go: downloading github.com/mattn/go-colorable v0.1.6
go: downloading github.com/sirupsen/logrus v1.5.0
go: downloading github.com/konsorten/go-windows-terminal-sequences v1.0.1

[nix-shell:~/tree/3rdparty/vehicle-command]$ go build ./...

```

At this point, instructions say to install binaries into your path. I didn't do this, I don't like installing experimental stuff into my path.

Generate public/private key:

```sh
[nix-shell:~/tree/3rdparty/vehicle-command]$ go run cmd/tesla-keygen/main.go -key-name="linuxpenguins.xyz" create  > public_key.pem
[nix-shell:~/tree/3rdparty/vehicle-command]$ go run cmd/tesla-keygen/main.go -key-name="linuxpenguins.xyz" create

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEG6PIK2TbMo5FNKQzC0zETBxi9r4f
gwzM7sDGKC2bMEGnWNLV4DXcC2/oyGUraovuXN1tGb1QpQFGL1yUk0Sk/Q==
-----END PUBLIC KEY-----

[nix-shell:~/tree/3rdparty/vehicle-command]$ go run cmd/tesla-keygen/main.go -key-name="linuxpenguins.xyz" create
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEG6PIK2TbMo5FNKQzC0zETBxi9r4f
gwzM7sDGKC2bMEGnWNLV4DXcC2/oyGUraovuXN1tGb1QpQFGL1yUk0Sk/Q==
-----END PUBLIC KEY-----

[nix-shell:~/tree/3rdparty/vehicle-command]$ go run cmd/tesla-keygen/main.go -key-name="linuxpenguins.xyz" create
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEG6PIK2TbMo5FNKQzC0zETBxi9r4f
gwzM7sDGKC2bMEGnWNLV4DXcC2/oyGUraovuXN1tGb1QpQFGL1yUk0Sk/Q==
-----END PUBLIC KEY-----

```

Notes:

* As per docs, it gives me the exact same public key every time.

* Command line parsing is somewhat strict, the "create" must come last. Otherwise will complain and not tell you why.

* Private key goes to seahorse on my Linux system, can be downloaded from there. Although it looks like the code really wants it in seahorse (could be a pain if trying to run from a docker container).

* Not really sure what key-name is for, guessing it is just a name for storing key in keyring, and its value is not actually important, as long as you use the same name every time.

Put key on my website at the appropriate path: https://linuxpenguins.xyz/.well-known/appspecific/com.tesla.3p.public-key.pem

Will need to follow these [instructions](https://developer.tesla.com/docs/fleet-api#register) to enroll, not done yet. Want to get the proxy running first. I don't want to enroll a public key, find I have to change it for some stupid reason, and then find out at that point that I can't change it.

Generate https key, think this is just a typical self signed https key. I might would end up putting this behind a https proxy that uses a signed certificate (but private access only of course!).

```sh
[nix-shell:~/tree/3rdparty/vehicle-command]$ openssl req -x509 -nodes -newkey ec -pkeyopt ec_paramgen_curve:secp521r1     -subj '/CN=localhost'     -keyout key.pem -out cert.pem -sha256 -days 3650     -addext "extendedKeyUsage = serverAuth"     -addext "keyUsage = digitalSignature, keyCertSign, keyAgreement"
```

Run server:

```sh
[nix-shell:~/tree/3rdparty/vehicle-command]$ go run ./cmd/tesla-http-proxy/main.go  -tls-key key.pem  -cert cert.pem -port 4443 -key-name "linuxpenguins.xyz"
Error: could not load key: The specified item could not be found in the keyring
exit status 1
[nix-shell:~/tree/3rdparty/vehicle-command]$ go run ./cmd/tesla-http-proxy/main.go  -tls-key key.pem  -cert cert.pem -port 4443 -key-name "vehicleCommandKey.linuxpenguins.xyz"
Error: could not load key: The specified item could not be found in the keyring
exit status 1
```

Huh? Obviously I am doing something (stupid?) wrong here, but can't figure it out right now.

**Comment by brianmay** [comment]
Relevant bug reports asking for protobuf stuff to be published:

* https://github.com/teslamotors/vehicle-command/issues/37
* https://github.com/teslamotors/vehicle-command/issues/72

**Comment by brianmay** [comment]
Filed a bug report on not being able to find the private key: https://github.com/teslamotors/vehicle-command/issues/73

**Comment by brianmay** [comment]
Looks like it isn't possible to enrol a certificate without a dev/partner account:

```
$ TESLA_API_TOKEN="$(pass show tesla/access_token)" curl --header 'Content-Type: application/json' \
  --header "Authorization: Bearer $TESLA_API_TOKEN" \
  --data '{"domain":"linuxpenguins.xyz"}' \
  'https://fleet-api.prd.na.vn.cloud.tesla.com/api/1/partner_accounts'

{"response":null,"error":"Expected partner account","error_description":""}%
```

Registering as a developer assumes you are a legal business.  https://developer.tesla.com/

![image](https://github.com/teslamate-org/teslamate/assets/112729/daebce23-d2aa-4304-981f-ed599e19b2e8)

![image](https://github.com/teslamate-org/teslamate/assets/112729/7b7c7b70-0e2f-4132-8093-672c7a8a26f7)

Somehow I am liking this less and less. It would appear that Tesla only want big business to use the API who can afford to spend $$$$.

And we need to go though this process to register the public keys for our vehicles.

**Reply by LelandSindt** [reply]
@brianmay 
Have you attempted the developer registration/request process? 

I have been un-successful so far... 

![image](https://github.com/teslamate-org/teslamate/assets/939903/ccc0b7d4-feae-4e4d-8665-5097971ffd03)

https://gaussmeter.bender.sway.org/ is serving certificates issued by Let's Encrypt... :shrug:

**Reply by brianmay** [reply]
I haven't got this far. A bit concerned with some of the terms in the [Fleet API agreement](https://www.tesla.com/legal/terms#fleet-api-agreement) at present. Wondering if any of these will potentially limit my ability to work on Teslamate.

For example, there is a clause that prohibits "Use the API for the development of a competing software product or service.". Which means if Tesla writes some sort of fleet management system, they could consider Teslamate a competitor, and if I agreed to this I would not be able to work on Teslamate. Unless I hire lawyers etc to argue the point it isn't really a competing service, different target audience, etc.

But note I am not a lawyer, nor have I consulted with one :-)

**Reply by eiannone** [reply]
I was able to register an account. Had to provide business information (including VAT id). The account was approved within minutes, and I got a client_id and a client secret.
I have successfully used them to obtain both a partner OAuth and a third party OAuth. The difference between the two types of tokens is not entirely clear to me.
I have installed the go SDK on a spare Raspberry Pi Zero, and obtained the public and private keys.
In the next days I will try to play with vehicle-command code. My plan is to develop my own library in Javascript, bypassing the proxy.

**Reply by eiannone** [reply]
@LelandSindt , did you try to use only the domain, instead of the full subdomain? I suggest using https://www.sway.org

**Reply by LelandSindt** [reply]
@eiannone I took some time to work on this today. I tried a few variations on `https://sub-domain.domain.[com|org]` and nothing worked until went to `https://domain.com` (note: I did not test `https://domain.org` ) 

additional notes: I am using let's encrypt for https/tls certs. The `domain.com` record that worked was an `A` record, all of the `sub-domain.domain.[com|org]` records were `cname` records.

**Comment by brianmay** [comment]
Currently the proxy http server is rejecting incoming requests with 403. Looking at the code, this is where it is failing:

```
func (p *oauthPayload) domain() string {
	for _, u := range p.Audiences {
		if strings.HasPrefix(u, "https://auth.tesla.com") {
			continue
		}
		d, _ := strings.CutPrefix(u, "https://")
		d, _ = strings.CutSuffix(d, "/")
		if !domainRegEx.MatchString(d) {
			continue
		}
		if remapped, ok := remappedDomains[d]; ok {
			return remapped
		}
		if strings.HasSuffix(d, ".tesla.com") || strings.HasSuffix(d, ".tesla.cn") {
			return d
		}
	}
	return ""
}

// New returns an [Account] that can be used to fetch a [vehicle.Vehicle].
func New(oauthToken string) (*Account, error) {
	parts := strings.Split(oauthToken, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("client provided malformed OAuth token")
	}
	payloadJSON, err := base64.RawStdEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("client provided malformed OAuth token: %s (%s)", err, parts[1])
	}
	var payload oauthPayload
	if err := json.Unmarshal(payloadJSON, &payload); err != nil {
		return nil, fmt.Errorf("client provided malformed OAuth token: %s", err)
	}

	domain := payload.domain()
	if domain == "" {
		return nil, fmt.Errorf("client provided OAuth token with invalid audiences")
	}
	return &Account{
		UserAgent:  UserAgent,
		authHeader: "Bearer " + strings.TrimSpace(oauthToken),
		Host:       domain,
	}, nil
}
```

Source: https://github.com/teslamotors/vehicle-command/blob/6f8d9adcceb71f3ef27ddc1b23a0f4da49250f36/pkg/account/account.go#L119

With debugging added, I am seeing the `client provided OAuth token with invalid audiences` message. Yes, could change the code, but strongly suspect that the Tesla server will make the same check.

So open question is, can we convince Tesla auth API to give as a token with the correct audience? Or is this something else you need a dev/partner account for?

**Comment by eiannone** [comment]
An alternative way to save the private key directly into file:
```
tesla-keygen -key-file ~/tesla_private.key -keyring-type file create
```
This will save the private key in clear, to `~/tesla_private.key` file.

**Comment by nelox** [comment]
teslamate.org or another entity should be the "Licensee" here. Where the submit application page mentions "business", it only means legal entity. As per https://www.tesla.com/legal/terms#fleet-api-agreement “Licensee” means the person or entity identified as the “Licensee” in the Key Terms.

**Comment by ogudet** [comment]
Helo,
I'm also trying to connect.
One thing did not understand is the usage of the private key.
As we put the public in "public" on the web site, and the fact the user transmit the public key to the car via the app. I don't see were we use the private key ?

**Comment by brianmay** [comment]
I imagine whenever you make a command request, it is signed using your private key, and then the car checks the signature of the request using the public key.

Meaning even if the security of the Tesla fleet API server was compromised, somebody would not be able to make requests to your car (at least not easily) without your private key.

Make sense?
